p2p/server.py

```diff
@@ -74,6 +77,56 @@ def __init__(self,
             min_peers=min_peers,
         )
 
+    async def refresh_nat_portmap(self) -> None:
+        """Run an infinite loop refreshing our NAT port mapping.
+
+        On every iteration we configure the port mapping with a lifetime of 30 minutes and then
+        sleep for that long as well.
+        """
+        lifetime = 30 * 60
+        while True:
+            self.logger.info("Setting up NAT portmap...")
+            # This is experimental and it's OK if it fails, hence the bare except.
+            try:
+                await self._add_nat_portmap(lifetime)
+            except Exception as e:
+                if (isinstance(e, upnpclient.soap.SOAPError) and
+                        e.args == (718, 'ConflictInMappingEntry')):
+                    # An entry already exists with the parameters we specified. Maybe the router
+                    # didn't clean it up after it expired or it has been configured by other piece
+                    # of software, either way we should not override it.
+                    # https://tools.ietf.org/id/draft-ietf-pcp-upnp-igd-interworking-07.html#errors
+                    self.logger.info("NAT port mapping already configured, not overriding it")
+                else:
+                    self.logger.exception("Failed to setup NAT portmap")
+
+            try:
+                await wait_with_token(asyncio.sleep(lifetime), token=self.cancel_token)
+            except OperationCancelled:
+                break
+
+    async def _add_nat_portmap(self, lifetime: int) -> None:
+        loop = asyncio.get_event_loop()
+        # Use loop.run_in_executor() because upnpclient.discover() is blocking and may take a
+        # while to complete.
+        devices = await wait_with_token(
+            loop.run_in_executor(None, upnpclient.discover),
+            token=self.cancel_token)
+        if not devices:
+            self.logger.info("No UPNP-enabled devices found")
+            return
+        device = devices[0]
+        device.WANIPConn1.AddPortMapping(
```

Comment:
No need to deal with this now, but I *think* that in my tinkering yesterday I found this property was not reliably present on all device objects which showed up in my network.

Reply:
Yeah, I'm sure this is going to crash in all sorts of ways, but I guess we'll have to rely on users reporting those crashes back to us so that we can handle all variants of this property